### PR TITLE
Fix Kotlin compiler warning rewriting when configured to turn warnings to errors

### PR DESCRIPTION
**Pull request by eskatos**
Fixes https://github.com/gradle/gradle/issues/15747

### CODE REVIEW

**Comment by eskatos** [comment]
@bot-gradle test this

File: subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslCompilerPlugins.kt [hunk 1]
```diff
@@ -96,7 +96,7 @@ private
 fun newLoggerMessageRewriterFor(experimentalWarning: Boolean, target: String, link: String) =
     { logLevel: LogLevel, message: String ->
         when {
-            logLevel != LogLevel.WARN -> message
+            logLevel != LogLevel.WARN && logLevel != LogLevel.ERROR -> message
```
Comment (bamboo) [inline subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslCompilerPlugins.kt:99]: 🤔 I wonder if we shouldn't just remove this optimization and let the `!message.contains(...)`  be the only condition behind the rewriting. We could also let the code return a specialized lambda depending on `experimentalWarning`. How about something like:

```kotlin
private
fun newLoggerMessageRewriterFor(experimentalWarning: Boolean, target: String, link: String) =
    when (experimentalWarning) {
        true -> { _: LogLevel, message: String ->
            when {
                isKotlinExperimentalWarning(message) -> kotlinDslPluginExperimentalWarning(target, link)
                else -> message
            }
        }
        false -> { _: LogLevel, message: String ->
            when {
                isKotlinExperimentalWarning(message) -> null
                else -> message
            }
        }
    }


private
fun isKotlinExperimentalWarning(message: String) =
    message.contains(KotlinCompilerArguments.samConversionForKotlinFunctions)
```

Comment (eskatos) [reply subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslCompilerPlugins.kt:99]: This code path is reached for each log statement, I would not remove the optimization.

Comment (bamboo) [reply subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslCompilerPlugins.kt:99]: Ok.

**Review by bamboo — APPROVED** [review]

**Pull request merged**